### ISSUE TITLE
C++20, Clang-10: Ambuiguity in resolving overloaded operator != in PartitionQosPolicy class [9825]

### ISSUE THREAD

**Issue by cvilas** [eProsima/Fast-DDS#1381]
FastDDS v2.0.1 is configured, built and installed as per README but with clang-10 and libc++. This step works fine. On attempting to build my own software (which uses fast DDS) with clang-10 and language set to `-std=c++2a`, I get the following compiler error:

```
In file included from /<install_path>/include/fastrtps/attributes/TopicAttributes.h:25:
In file included from /<install_path>/include/fastrtps/qos/QosPolicies.h:23:
/<install_path>/include/fastdds/dds/core/policy/QosPolicies.hpp:1499:40: error: use of overloaded operator '!=' is ambiguous (with operand types 'eprosima::fastdds::dds::PartitionQosPolicy::const_iterator' and 'eprosima::fastdds::dds::PartitionQosPolicy::const_iterator')
            for (auto it = begin(); it != end(); ++it)
                                    ~~ ^  ~~~~~
/<install_path>/include/fastdds/dds/core/policy/QosPolicies.hpp:1270:14: note: candidate function
        bool operator !=(
             ^
/<install_path>/include/fastdds/dds/core/policy/QosPolicies.hpp:1264:14: note: candidate function
        bool operator ==(
             ^
/<install_path>/include/fastdds/dds/core/policy/QosPolicies.hpp:1264:14: note: candidate function (with reversed parameter order)
``` 

## Expected Behavior
Able to use fastDDS with client code compiled for C++20.

## Current Behavior
When compiling my code for C++20 (i.e. its CMakeLists.txt has `set(CMAKE_CXX_STANDARD 20)`) using FastDDS installed as per previous step, the aforementioned compile error is the result. Everything works fine with gcc and libstdc++

## Steps to Reproduce
* FastDDS was built and installed with `CC=clang CXX="clang++ -stdlib=libc++" cmake <other options> && make && make install`
* My code was being built the same way and has `set(CMAKE_CXX_STANDARD 20)` in it's top level CMakeLists.txt
* I don't have an example program to illustrate. Apologies.

## System information
- Fast-RTPS version: v2.0.1 tar.gz downloaded from here
- OS: Ubuntu 20.04
- Compiler: Clang-10
- Network interfaces: irrelevant
- ROS2: No

## Additional context
A similar issue is described here: https://stackoverflow.com/questions/60568088/breaking-change-in-c20-or-regression-in-clang-trunk-gcc-trunk-when-overloading

As mentioned in the reply with the most votes, the fix is straightforward and involves making the overloads  `PartitionQosPolicy::operator==()` and `PartitionQosPolicy::operator!=()` in lines 1264 and 1270 of QosPolicies.hpp, respectively, const

```
bool operator ==(const self_type& rhs) const;
bool operator !=(const self_type& rhs) const;
                                       ^^^^^^
```

Patch:
```
@@ -1262,13 +1262,13 @@
         }
 
         bool operator ==(
-                const self_type& rhs)
+                const self_type& rhs) const
         {
             return ptr_ == rhs.ptr_;
         }
 
         bool operator !=(
-                const self_type& rhs)
+                const self_type& rhs) const
         {
             return ptr_ != rhs.ptr_;
         }

```
## Additional resources
none

**Comment by cvilas** [comment]
Is there a timeline for when this fix might go into `master`?

**Comment by JaimeMartin** [comment]
Hi @cvilas 

Today is a local holiday here, we will answer you this week. @imontesino, @MiguelCompany, please give @cvilas an answer on this.

**Comment by MiguelCompany** [comment]
This should have been solved by #1634.
